File: dox_development/dox_development/report/customer_statement/customer_statement.js
// Copyright (c) 2023, Nesscale Solutions Private Limited and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Customer Statement"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"width": "80",
			"options": "Company",
			"default": frappe.defaults.get_default("company"),
			"hidden":true
		},
		{
			"fieldname": "party",
			"label": __("Party"),
			"fieldtype": "Link",
			"options": "Customer",
			"reqd": 1,
			get_data: function (txt) {
				if (!frappe.query_report.filters) return;

				let party_type = frappe.query_report.get_filter_value('party_type');
				if (!party_type) return;

				return frappe.db.get_link_options(party_type, txt);
			},
			on_change: function () {
				var party_type = "Customer";
				var parties = frappe.query_report.get_filter_value('party');
				var party = parties;
				var fieldname = erpnext.utils.get_party_name(party_type) || "name";
				frappe.db.get_value(party_type, party, fieldname, function (value) {
					frappe.query_report.set_filter_value('party_name', value[fieldname]);
				});
				// if (!party_type || parties.length === 0 || parties.length > 1) {
				// 	frappe.query_report.set_filter_value('party_name', "");
				// 	return;
				// } else {

				// }
			}
		},
		{
			"fieldname": "party_name",
			"label": __("Party Name"),
			"fieldtype": "Read Only"
		},
		{
			"fieldname": "branch",
			"label": __("Branch"),
			"fieldtype": "Link",
			"options": "Branch",
			"get_query": function() {
				// Get the report data
				var reportData = frappe.query_report.data;
		
				// Get the unique branches from the report data
				var reportBranches = reportData.map(row => row.branch);
				var uniqueBranches = [...new Set(reportBranches)];
		
				return {
					filters: {
						name: ["in", uniqueBranches]
					}
				};
			}
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1,
			"width": "60px"
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1,
			"width": "60px"
		},
		{
			"fieldname": "address",
			"label": __("Address"),
			"fieldtype": "Link",
			"options": "Address",
			"get_query": function () {
				var party = frappe.query_report.get_filter_value('party');
				if (party) {
					return {
						filters: {
							link_doctype: "Customer",
							link_name: party
						}
					};
				}
			}
		},
		{
			"fieldname": "remark",
			"label": __("Remark"),
			"fieldtype": "Check"
		},
		{
			"fieldname": "terms",
			"label": __("Terms and Conditons"),
			"fieldtype": "Link",
			"options":"Terms and Conditions",
			"width": "60px"
		},
	],
    onload: function(report) {
        // Listen for changes in 'filter_1'
        report.page.fields_dict.party.$input.on('change', function() {
            // Clear values of 'filter_2' and 'filter_3' when 'filter_1' changes
            report.set_filter_value('branch', '');
            report.set_filter_value('address', '');
        });
    }
};

File: dox_development/dox_development/report/customer_statement/customer_statement.py
import frappe
from frappe import _


def execute(filters=None):
    frappe.errprint("call")
    columns = []
    if not filters.get("from_date") or not filters.get("to_date"):
        frappe.throw(_("Select First from date and to date"))
        return columns, []

    if not filters.get("party"):
        frappe.throw(_("Select Customer to view statement"))
        return columns, []

    filters_report = {
        "company": filters.get("company"),
        "from_date": filters.get("from_date"),
        "to_date": filters.get("to_date"),
        "party_type": "Customer",
        "party": [filters.get("party")],
        "party_name": filters.get("party_name"),
        "group_by": "Group by Voucher (Consolidated)",
        "branch": [] if not filters.get("branch") else [filters.get("branch")],
        "cost_center": [],
        "project": [],
        "include_dimensions": 1,
    }

    from frappe.desk.query_report import run, add_custom_column_data

    res = run("General Ledger", filters=filters_report)
    data = res.get("result")

    columns = [
        {
            "label": "Posting Date",
            "fieldname": "posting_date",
            "fieldtype": "Date",
            "width": 120,
        },
        {"label": "Voucher Type", "fieldname": "voucher_type", "width": 140},
        {
            "label": "Voucher No",
            "fieldname": "voucher_no",
            "fieldtype": "Dynamic Link",
            "options": "voucher_type",
            "width": 180,
        },
        {
            "label": "Branch",
            "options": "Branch",
            "fieldname": "branch",
            "width": 100,
            "hidden": 1,
        },
        {
            "label": "Debit (SAR)",
            "fieldname": "debit",
            "fieldtype": "Float",
            "width": 120,
        },
        {
            "label": "Credit (SAR)",
            "fieldname": "credit",
            "fieldtype": "Float",
            "width": 120,
        },
        {
            "label": "Balance (SAR)",
            "fieldname": "balance",
            "fieldtype": "Float",
            "width": 130,
        },
    ]
    if filters.remark:
        columns.insert(
            3,
            {
                "label": _("Remarks"),
                "fieldname": "remarks",
                "insert_after_index": 1,
                "width": 400,
            },
        )

    custom_columns = [
        {
            "doctype": "Payment Entry",
            "fieldname": "mode_of_payment",
            "link_field": "voucher_no",
        },
        {
            "doctype": "Sales Invoice",
            "fieldname": "invoice_type",
            "link_field": "voucher_no",
        },
        {
            "doctype": "Payment Entry",
            "fieldname": "payment_type",
            "link_field": "voucher_no",
        },
    ]

    add_custom_column_data(custom_columns, data)

    for row in data:
        if row.get("voucher_type") == "Sales Invoice":
            if row.get("invoice_type") == "Credit Note":
                row["voucher_type"] = "Sales Return"
            else:
                row["voucher_type"] = row.get("invoice_type")
        if row.get("voucher_type") == "Payment Entry":
            row[
                "voucher_type"
            ] = f"{row.get('payment_type')} / {row.get('mode_of_payment')}"
        if row.get("voucher_type") == "Journal Entry":
            row["voucher_type"] = "Adjustment Entry"

        address = frappe.db.get_value(
            "Address",
            {"name": filters.get("address")},
            ["address_title", "address_line1", "city", "county", "country"],
            as_dict=1,
        )
        if address:
            row["address"] = address
        else:
            row["address"] = {}

        terms = frappe.db.get_value(
            "Terms and Conditions", filters.get("terms"), "terms", as_dict=1
        )
        if terms:
            row["terms"] = terms.get("terms")
        else:
            row["terms"] = {}
    return columns, data
